zeus/profiling.py

```diff
@@ -0,0 +1,239 @@
+# Copyright (C) 2022 Jae-Won Chung <jwnchung@umich.edu>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for profiling energy and time inside training script."""
+
+from __future__ import annotations
+
+import time
+import subprocess
+import signal
+import atexit
+import logging
+from typing import List, Union, Tuple
+
+import pynvml
+
+from zeus import analyze
+from zeus.util.logging import LOG
+
+
+class ZeusProfilingService:
+    """Profiling the energy and time inside training script.
+
+    Push a profile to start profiling and pop a profile to get the profiling result
+    of the *last* pushed profiling window. Multiple profiling windows are stored
+    in a stack and conform to *First-In-First-Out*.
+
+    ## Integrated Example
+    ```python
+    prof_service = zeus.profiling.ZeusProfilingService(gpu_handles, monitor_path, power_log_prefix)
+
+    # Push a profile window
+    prof_service.push_window()
+
+    # Actual work
+    training(x, y)
+
+    # Pop a profile window and get the profiling result
+    time_consumed, energy_consumed = prof_service.pop_window()
+
+    print(f"Training takes {time_consumed} seconds.")
+    for gpu_idx, energy_consumed_per_gpu in energy_consumed:
+        print(f"GPU {gpu_idx} consumes {energy_consumed_per_gpu} Joules.")
+    ```
+    Please checkout out [zeus.run.dataloader] for a complete integrated example.
+    """
+
+    def __init__(
+        self,
+        gpu_handles: List[pynvml.c_nvmlDevice_t],
```

Comment:
It's fine if multiple handles exist in the same program. Also, if you weren't aware, `nvmlInit` can also be called multiple times. NVML will just keep a stack internally, and only when the same number of `nvmlShutdown` is called it will truely shut down. So I don't think you'll even need `get_gpu_handles()` because nobody will really need it. The handle instantiated inside the class can be kept private.